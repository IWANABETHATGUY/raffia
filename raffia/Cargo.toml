[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Generic parser for CSS/SCSS/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_derive = { path = "../raffia_derive" }
