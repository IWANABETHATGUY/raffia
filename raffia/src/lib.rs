pub use config::Syntax;
pub use parser::{parse, parse_with_comments, Parse, Parser};
pub use pos::{Span, Spanned};

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod tokenizer;
mod util;
