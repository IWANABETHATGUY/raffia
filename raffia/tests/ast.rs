use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = parser.parse::<Stylesheet>().unwrap();

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
