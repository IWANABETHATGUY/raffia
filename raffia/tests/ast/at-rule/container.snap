---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "container",
        raw: "container",
        span: Span(
          start: 1,
          end: 10,
        ),
      ),
      prelude: Some(ContainerPrelude(
        type: "ContainerPrelude",
        name: Some(Ident(
          type: "Ident",
          name: "name",
          raw: "name",
          span: Span(
            start: 11,
            end: 15,
          ),
        )),
        condition: ContainerCondition(
          type: "ContainerCondition",
          conditions: [
            MediaFeatureBoolean(
              type: "MediaFeatureBoolean",
              name: Ident(
                type: "Ident",
                name: "width",
                raw: "width",
                span: Span(
                  start: 17,
                  end: 22,
                ),
              ),
              span: Span(
                start: 17,
                end: 22,
              ),
            ),
          ],
          span: Span(
            start: 17,
            end: 22,
          ),
        ),
        span: Span(
          start: 11,
          end: 22,
        ),
      )),
      block: Some(SimpleBlock(
        type: "SimpleBlock",
        statements: [],
        span: Span(
          start: 24,
          end: 26,
        ),
      )),
      span: Span(
        start: 0,
        end: 26,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "container",
        raw: "container",
        span: Span(
          start: 28,
          end: 37,
        ),
      ),
      prelude: Some(ContainerPrelude(
        type: "ContainerPrelude",
        name: Some(Ident(
          type: "Ident",
          name: "name",
          raw: "name",
          span: Span(
            start: 38,
            end: 42,
          ),
        )),
        condition: ContainerCondition(
          type: "ContainerCondition",
          conditions: [
            ContainerCondition(
              type: "ContainerCondition",
              conditions: [
                MediaFeatureBoolean(
                  type: "MediaFeatureBoolean",
                  name: Ident(
                    type: "Ident",
                    name: "width",
                    raw: "width",
                    span: Span(
                      start: 45,
                      end: 50,
                    ),
                  ),
                  span: Span(
                    start: 45,
                    end: 50,
                  ),
                ),
              ],
              span: Span(
                start: 45,
                end: 50,
              ),
            ),
            ContainerConditionOr(
              type: "ContainerConditionOr",
              keyword: Ident(
                type: "Ident",
                name: "or",
                raw: "or",
                span: Span(
                  start: 53,
                  end: 55,
                ),
              ),
              queryInParens: ContainerCondition(
                type: "ContainerCondition",
                conditions: [
                  ContainerConditionNot(
                    type: "ContainerConditionNot",
                    keyword: Ident(
                      type: "Ident",
                      name: "not",
                      raw: "not",
                      span: Span(
                        start: 57,
                        end: 60,
                      ),
                    ),
                    queryInParens: ContainerCondition(
                      type: "ContainerCondition",
                      conditions: [
                        MediaFeatureBoolean(
                          type: "MediaFeatureBoolean",
                          name: Ident(
                            type: "Ident",
                            name: "width",
                            raw: "width",
                            span: Span(
                              start: 63,
                              end: 68,
                            ),
                          ),
                          span: Span(
                            start: 63,
                            end: 68,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 63,
                        end: 68,
                      ),
                    ),
                    span: Span(
                      start: 57,
                      end: 68,
                    ),
                  ),
                ],
                span: Span(
                  start: 57,
                  end: 68,
                ),
              ),
              span: Span(
                start: 53,
                end: 68,
              ),
            ),
          ],
          span: Span(
            start: 45,
            end: 68,
          ),
        ),
        span: Span(
          start: 38,
          end: 68,
        ),
      )),
      block: Some(SimpleBlock(
        type: "SimpleBlock",
        statements: [
          QualifiedRule(
            type: "QualifiedRule",
            selector: SelectorList(
              type: "SelectorList",
              selectors: [
                ComplexSelector(
                  type: "ComplexSelector",
                  children: [
                    CompoundSelector(
                      type: "CompoundSelector",
                      children: [
                        TagNameSelector(
                          type: "TagNameSelector",
                          name: WqName(
                            type: "WqName",
                            name: Ident(
                              type: "Ident",
                              name: "main",
                              raw: "main",
                              span: Span(
                                start: 74,
                                end: 78,
                              ),
                            ),
                            prefix: None,
                            span: Span(
                              start: 74,
                              end: 78,
                            ),
                          ),
                          span: Span(
                            start: 74,
                            end: 78,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 74,
                        end: 78,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 74,
                    end: 78,
                  ),
                ),
              ],
              span: Span(
                start: 74,
                end: 78,
              ),
            ),
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "--match",
                    raw: "--match",
                    span: Span(
                      start: 81,
                      end: 88,
                    ),
                  ),
                  value: [
                    Ident(Ident(
                      kind: "Ident",
                      name: "true",
                      raw: "true",
                      span: Span(
                        start: 89,
                        end: 93,
                      ),
                    )),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 81,
                    end: 93,
                  ),
                ),
              ],
              span: Span(
                start: 79,
                end: 96,
              ),
            ),
            span: Span(
              start: 74,
              end: 96,
            ),
          ),
        ],
        span: Span(
          start: 72,
          end: 98,
        ),
      )),
      span: Span(
        start: 27,
        end: 98,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 99,
  ),
)
